File: 0039. Combination Sum/Solution.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        findCombinationSum(0,target,res,candidates,new ArrayList<>());
        return res;
    }

    private void findCombinationSum(int index,int target,List<List<Integer>> res,int[] arr,List<Integer> curr){
        if(index==arr.length){
            if(target==0){
                res.add(new ArrayList<>(curr));
            }
            return;
        }

        if(arr[index]<=target){
            curr.add(arr[index]);
            findCombinationSum(index,target-arr[index],res,arr,curr);
            curr.remove(curr.size() - 1);
        }
        findCombinationSum(index+1,target,res,arr,curr);
    }
}

File: 0039. Combination Sum/Solution.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        self.helper(0,target,res,[],candidates)
        return res
        

    def helper(self,index,target,res,curr,nums):
        if(index==len(nums)):
            if(target==0):
                res.append(curr[:])
            return

        if(target>=nums[index]):
            curr.append(nums[index])
            self.helper(index,target-nums[index],res,curr,nums)
            curr.pop()

        self.helper(index+1,target,res,curr,nums)

